File: fixers/copyright-continued-lines-with-space.py
#!/usr/bin/python3

import re
import sys
from typing import List

from lintian_brush.fixer import LintianIssue, report_result


def expand_tabs(line, tabwidth=8):
    ret: List[bytes] = []
    for i, _c in enumerate(line):
        if line[i:i+1] == b'\t':
            # Round up to the next unit of tabwidth
            ret.extend([b' '] * (tabwidth - len(ret) % tabwidth))
        else:
            ret.append(line[i:i+1])
    return b''.join(ret)


def whitespace_prefix_length(line):
    m = re.match(b'^\\s*', line)
    if not m:
        return 0
    return len(m.group(0))


def value_offset(line):
    if not line.strip():
        return None
    if line.startswith(b'#'):
        return None
    if line[0:1] in (b'\t', b' '):
        return whitespace_prefix_length(line)
    else:
        try:
            key, value = line.split(b':', 1)
        except ValueError:
            return None
        else:
            return len(key) + 1 + whitespace_prefix_length(value)


EXPECTED_HEADER = (
    b'Format: '
    b'https://www.debian.org/doc/packaging-manuals/copyright-format/1.0')


UNICODE_LINE_BREAK = "\u2028".encode()
UNICODE_PARAGRAPH_SEPARATOR = "\u2029".encode()


tabs_replaced = False
unicode_linebreaks_replaced = False

try:
    lines = []
    with open('debian/copyright', 'rb') as f:
        line = f.readline()
        if line.rstrip().rstrip(b'/') != EXPECTED_HEADER:
            # Not a machine-readable copyright file
            sys.exit(0)
        lines.append(line)
        prev_value_offset = None
        for lineno, line in enumerate(f, start=2):
            if line.startswith(b'\t'):
                issue = LintianIssue(
                    'source', 'tab-in-license-text',
                    info='debian/copyright (paragraph at line %d)' % lineno)
                if issue.should_fix():
                    for option in [
                            b' \t' + line[1:], b' \t' + line[2:],
                            b' ' * 8 + line[1:]]:
                        if value_offset(option) == prev_value_offset:
                            line = option
                            break
                    else:
                        line = b' \t' + line[1:]
                    tabs_replaced = True
                    issue.report_fixed()
            if UNICODE_PARAGRAPH_SEPARATOR in line:
                # Not quite the same thing, but close enough..
                line = line.replace(
                    UNICODE_PARAGRAPH_SEPARATOR,
                    UNICODE_LINE_BREAK * 2)
            if UNICODE_LINE_BREAK in line:
                unicode_linebreaks_replaced = True
                parts = line.split(UNICODE_LINE_BREAK)
                # If the line is empty, replace it with a .
                parts = [p if p else b'.' for p in parts]
                parts = [parts[0]] + [b' ' + p for p in parts[1:]]
                line = b'\n'.join(parts)
            lines.append(line)
            prev_value_offset = value_offset(line)
except FileNotFoundError:
    pass
else:
    with open('debian/copyright', 'wb') as g:
        g.writelines(lines)

sys.stdout.write('debian/copyright: ')
if tabs_replaced:
    sys.stdout.write('use spaces rather than tabs to start continuation lines')
    if unicode_linebreaks_replaced:
        sys.stdout.write(', ')
if unicode_linebreaks_replaced:
    sys.stdout.write('replace unicode linebreaks with regular linebreaks')
sys.stdout.write('.\n')
report_result()

File: lintian-brush/src/fixers/copyright_continued_lines_with_space.rs
use crate::{declare_fixer, FixerError, FixerResult};
use std::fs;
use std::path::Path;

const EXPECTED_HEADER: &[u8] =
    b"Format: https://www.debian.org/doc/packaging-manuals/copyright-format/1.0";
const UNICODE_LINE_BREAK: &[u8] = "\u{2028}".as_bytes();
const UNICODE_PARAGRAPH_SEPARATOR: &[u8] = "\u{2029}".as_bytes();

fn is_whitespace(b: u8) -> bool {
    b == b' ' || b == b'\t' || b == b'\n' || b == b'\r'
}

fn whitespace_prefix_length(line: &[u8]) -> usize {
    line.iter()
        .take_while(|&&b| b == b' ' || b == b'\t')
        .count()
}

fn value_offset(line: &[u8]) -> Option<usize> {
    if line.iter().all(|&b| is_whitespace(b)) {
        return None;
    }
    if line.starts_with(b"#") {
        return None;
    }
    if line.starts_with(b"\t") || line.starts_with(b" ") {
        return Some(whitespace_prefix_length(line));
    }
    // Look for key: value
    line.iter()
        .position(|&b| b == b':')
        .map(|colon_pos| colon_pos + 1 + whitespace_prefix_length(&line[colon_pos + 1..]))
}

fn split_bytes(data: &[u8], separator: &[u8]) -> Vec<Vec<u8>> {
    let mut result = Vec::new();
    let mut current = Vec::new();
    let mut i = 0;

    while i < data.len() {
        if i + separator.len() <= data.len() && &data[i..i + separator.len()] == separator {
            result.push(current);
            current = Vec::new();
            i += separator.len();
        } else {
            current.push(data[i]);
            i += 1;
        }
    }
    result.push(current);
    result
}

fn join_bytes(parts: Vec<Vec<u8>>, separator: &[u8]) -> Vec<u8> {
    let mut result = Vec::new();
    for (i, part) in parts.into_iter().enumerate() {
        if i > 0 {
            result.extend_from_slice(separator);
        }
        result.extend(part);
    }
    result
}

pub fn run(base_path: &Path) -> Result<FixerResult, FixerError> {
    let copyright_path = base_path.join("debian/copyright");

    if !copyright_path.exists() {
        return Err(FixerError::NoChanges);
    }

    let content = fs::read(&copyright_path)?;
    let mut lines = content.split_inclusive(|&b| b == b'\n').peekable();

    // Check the first line for the expected header
    let first_line = match lines.peek() {
        Some(&line) => line,
        None => return Err(FixerError::NoChanges),
    };

    // Strip whitespace, then trailing slashes (like Python's .rstrip().rstrip(b"/"))
    let mut trimmed = first_line;
    while trimmed.last().map_or(false, |&b| is_whitespace(b)) {
        trimmed = &trimmed[..trimmed.len() - 1];
    }
    while trimmed.last() == Some(&b'/') {
        trimmed = &trimmed[..trimmed.len() - 1];
    }

    if trimmed != EXPECTED_HEADER {
        return Err(FixerError::NoChanges);
    }

    let mut new_lines = Vec::new();
    let mut tabs_replaced = false;
    let mut unicode_linebreaks_replaced = false;
    let mut prev_value_offset: Option<usize> = None;

    for line in lines {
        let mut line = line.to_vec();

        // Handle tabs at the start of continuation lines
        if line.starts_with(b"\t") {
            // Try different replacement options to maintain alignment
            let make_option = |prefix: &[u8], skip: usize| {
                let mut v = prefix.to_vec();
                if line.len() > skip {
                    v.extend_from_slice(&line[skip..]);
                }
                v
            };

            let options = [
                make_option(&[b' ', b'\t'], 1),
                make_option(&[b' ', b'\t'], 2),
                make_option(&[b' '; 8], 1),
            ];

            line = options
                .into_iter()
                .find(|opt| value_offset(opt) == prev_value_offset)
                .unwrap_or_else(|| make_option(&[b' ', b'\t'], 1));

            tabs_replaced = true;
        }

        // Handle unicode paragraph separator (replace with two line breaks)
        if line
            .windows(UNICODE_PARAGRAPH_SEPARATOR.len())
            .any(|w| w == UNICODE_PARAGRAPH_SEPARATOR)
        {
            let parts = split_bytes(&line, UNICODE_PARAGRAPH_SEPARATOR);
            let separator = [UNICODE_LINE_BREAK, UNICODE_LINE_BREAK].concat();
            line = join_bytes(parts, &separator);
        }

        // Handle unicode line breaks
        if line
            .windows(UNICODE_LINE_BREAK.len())
            .any(|w| w == UNICODE_LINE_BREAK)
        {
            unicode_linebreaks_replaced = true;
            let parts = split_bytes(&line, UNICODE_LINE_BREAK);

            let new_parts: Vec<_> = parts
                .into_iter()
                .enumerate()
                .map(|(i, part)| {
                    let content = if part.is_empty() { b"." } else { &part[..] };
                    if i == 0 {
                        content.to_vec()
                    } else {
                        [&[b' '], content].concat()
                    }
                })
                .collect();

            line = join_bytes(new_parts, b"\n");
        }

        prev_value_offset = value_offset(&line);
        new_lines.push(line);
    }

    if !tabs_replaced && !unicode_linebreaks_replaced {
        return Err(FixerError::NoChanges);
    }

    let output: Vec<u8> = new_lines.into_iter().flatten().collect();
    fs::write(&copyright_path, output)?;

    let mut description = "debian/copyright: ".to_string();
    if tabs_replaced {
        description.push_str("use spaces rather than tabs to start continuation lines");
        if unicode_linebreaks_replaced {
            description.push_str(", ");
        }
    }
    if unicode_linebreaks_replaced {
        description.push_str("replace unicode linebreaks with regular linebreaks");
    }
    description.push('.');

    let mut result = FixerResult::builder(description);
    if tabs_replaced {
        result = result.fixed_tag("tab-in-license-text");
    }
    Ok(result.build())
}

declare_fixer! {
    name: "copyright-continued-lines-with-space",
    tags: ["tab-in-license-text"],
    apply: |basedir, _package, _version, _preferences| {
        run(basedir)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;
    use tempfile::TempDir;

    #[test]
    fn test_replace_tabs() {
        let temp_dir = TempDir::new().unwrap();
        let base_path = temp_dir.path();
        let debian_dir = base_path.join("debian");
        fs::create_dir(&debian_dir).unwrap();

        let copyright_path = debian_dir.join("copyright");
        fs::write(
            &copyright_path,
            b"Format: https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/\n\
License: GPL-3+\n\
\tThis is a continuation line\n",
        )
        .unwrap();

        let result = run(base_path).unwrap();
        assert!(result
            .description
            .contains("use spaces rather than tabs to start continuation lines"));

        let content = fs::read(&copyright_path).unwrap();
        let expected = b"Format: https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/\nLicense: GPL-3+\n \tThis is a continuation line\n";
        assert_eq!(content, expected);
    }

    #[test]
    fn test_no_changes() {
        let temp_dir = TempDir::new().unwrap();
        let base_path = temp_dir.path();
        let debian_dir = base_path.join("debian");
        fs::create_dir(&debian_dir).unwrap();

        let copyright_path = debian_dir.join("copyright");
        fs::write(
            &copyright_path,
            b"Format: https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/\n\
License: GPL-3+\n\
 This is a continuation line\n",
        )
        .unwrap();

        let result = run(base_path);
        assert!(matches!(result, Err(FixerError::NoChanges)));
    }

    #[test]
    fn test_not_machine_readable() {
        let temp_dir = TempDir::new().unwrap();
        let base_path = temp_dir.path();
        let debian_dir = base_path.join("debian");
        fs::create_dir(&debian_dir).unwrap();

        let copyright_path = debian_dir.join("copyright");
        fs::write(
            &copyright_path,
            b"This is a regular copyright file\n\
Copyright (c) 2024 Someone\n",
        )
        .unwrap();

        let result = run(base_path);
        assert!(matches!(result, Err(FixerError::NoChanges)));
    }

    #[test]
    fn test_unicode_linebreaks() {
        let temp_dir = TempDir::new().unwrap();
        let base_path = temp_dir.path();
        let debian_dir = base_path.join("debian");
        fs::create_dir(&debian_dir).unwrap();

        let copyright_path = debian_dir.join("copyright");
        let mut content = b"Format: https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/\nLicense: GPL-3+\n Line one".to_vec();
        content.extend_from_slice(UNICODE_LINE_BREAK);
        content.extend_from_slice(b"Line two\n");
        fs::write(&copyright_path, &content).unwrap();

        let result = run(base_path).unwrap();
        assert!(result
            .description
            .contains("replace unicode linebreaks with regular linebreaks"));

        let new_content = fs::read(&copyright_path).unwrap();
        let expected = b"Format: https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/\nLicense: GPL-3+\n Line one\n Line two\n";
        assert_eq!(new_content, expected);
    }

    #[test]
    fn test_no_copyright_file() {
        let temp_dir = TempDir::new().unwrap();
        let base_path = temp_dir.path();

        let result = run(base_path);
        assert!(matches!(result, Err(FixerError::NoChanges)));
    }
}
